enterprise/models.py

```diff
@@ -478,9 +563,14 @@ def enroll(self, course_run_id, mode):
         """
         enrollment_api_client = EnrollmentApiClient()
         # Check to see if the user's already enrolled and we have an enterprise course enrollment to track it.
-        enrolled_in_course = enrollment_api_client.is_enrolled(self.username, course_run_id)
-        if not enrolled_in_course:
-            # Directly enroll into the audit track.
+        course_enrollment = enrollment_api_client.get_course_enrollment(self.username, course_run_id) or {}
+        enrolled_in_course = course_enrollment and course_enrollment.get('is_active', False)
+
+        audit_modes = getattr(settings, 'ENTERPRISE_COURSE_ENROLLMENT_AUDIT_MODES', ['audit', 'honor'])
+        paid_modes = ['verified', 'professional']
+        is_upgrading = mode in paid_modes and course_enrollment.get('mode') in audit_modes
+        if not enrolled_in_course or is_upgrading:
```

Comment:
I think we probably want to modify the `edx.bi.user.enterprise.enrollment.course` tracking event below to identify mode changes, i.e. old mode vs. new mode.